### ISSUE TITLE
Fail to reproduce results on GSM8k

### ISSUE THREAD

**Issue by ThisisBillhe** [facebookresearch/coconut#38]
Hi, thanks for your great work!

I try to reproduce your results by following your guide and training coconut for 25 epochs. This is the val accuracy during training:

![Image](https://github.com/user-attachments/assets/d414ed5a-56c0-402e-a67f-cf3047682004)

I only get an accuracy of 31.13% after training, while the results reported in paper is 34.1%. Do you have any opinion?

**Comment by Ber666** [comment]
Hi, could you share the training config file you are using? The validation accuracy curve does look different from what I get. For example, at epoch 13 the accuracy is below 0.2. In my runs, the accuracy always stays above 0.26 and would finally coverage to around 0.35. It would also be helpful if you could share the training loss curve.

**Comment by ThisisBillhe** [comment]
I see. Thanks for your reply. I will try a smaller global batch size by setting batch_size_training to 16.

**Comment by Devansh1105** [comment]
In the paper you have mentioned that you train the coconut model for 50 epochs in total for gsm8k and where 6 epochs in initial stage that is CoT followed by 3 stages for 3 epochs each and then final stage where all reasoning tokens are removed and the tokens according to previous latent stage is used.

Doubts here:
1) Is the base model gpt2 used or is gpt2 trained on CoT used for above setup?
2) The github config is a bit different, with num epochs set to 25.
3) gsm_coconut.yaml has resume : 3 and seems to resume from somewhere, please clarify this as well.
4) Can we use the exact configs uploaded in the github or are tweaks required to replicate?

Can you please clarify as I am trying to replicate the results.

**Comment by Ber666** [comment]
Hi @Devansh1105, thanks for the question.
 
1. In our code, the initial stage training of coconut is replaced with CoT SFT. They are essentially equivalent, except for the `<bot>` and `<eot>` tokens, which don't really matter in terms of performance.
2. That's the max epochs for training. You should select the checkpoint based on validation set accuracy.
3. That's a trick to skip the initial stage training of coconut. Because we have loaded the CoT SFT checkpoint.
4. Yes, the exact config can reproduce the results.
